presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PayloadJoinOptimizer.java

```diff
@@ -192,16 +186,31 @@ public PlanNode visitUnion(UnionNode unionNode, RewriteContext<JoinContext> cont
             return context.defaultRewrite(unionNode, new JoinContext());
         }
 
+        @Override
+        public PlanNode visitAggregation(AggregationNode aggNode, RewriteContext<JoinContext> context)
+        {
+            if (aggNode.getOutputVariables().containsAll(context.get().getJoinKeys())) {
+                return context.defaultRewrite(aggNode, context.get());
+            }
+            return context.defaultRewrite(aggNode, new JoinContext());
+        }
+
         @Override
         public PlanNode visitJoin(JoinNode joinNode, RewriteContext<JoinContext> context)
         {
             final JoinContext joinContext = context.get();
-            int numJoinKeys = joinContext.getJoinKeys().size();
+            Set<VariableReferenceExpression> oldJoinKeys = joinContext.getJoinKeys();
             PlanNode leftNode = joinNode.getLeft();
             PlanNode rightNode = joinNode.getRight();
-            boolean isTopJoin = numJoinKeys == 0;
+
+            boolean isTopJoin = joinContext.getJoinKeys().size() == 0;
 
             ImmutableSet<VariableReferenceExpression> leftColumns = leftNode.getOutputVariables().stream().collect(toImmutableSet());
+
+            // remove join keys coming from RHS of join
+            ImmutableSet<VariableReferenceExpression> redundantJoinKeys = oldJoinKeys.stream().filter(key -> rightNode.getOutputVariables().contains(key)).collect(toImmutableSet());
```

Comment:
calling these redundant makes this sound like an optimization. but as i understand it's a correctness/valid plan issue.